PHI-base is an online database (available at
[phi-base.org](http://www.phi-base.org)) that catalogues experimentally
verified pathogenicity, virulence and effector genes from fungal,
oomycete and bacterial pathogens, which infect animal, plant, fungal and
insect hosts. PHI-base is a valuable resource in the discovery of genes
in medically and agronomically important pathogens, which may be
potential targets for chemical intervention.

Each entry in PHI-base is curated by domain experts and is supported by
strong experimental evidence (for example, gene disruption and gene
complementation experiments), as well as literature references in which
the original experiments are described. Each gene in PHI-base is
presented with its nucleotide sequence and deduced amino acid sequence
(available in a FASTA file), as well as a detailed description of the
predicted protein's function during the host infection process. To
facilitate data interoperability, we have annotated genes using
ontologies, controlled vocabularies, and links to external sources
(including UniProt, Gene Ontology, Enzyme Commission, NCBI Taxonomy,
EMBL, PubMed and FRAC).

This PHI-base dataset is a Frictionless Data Package that contains an
export of the PHI-base database in CSV format (comma-separated values),
plus a FASTA file with sequences for each gene in the database. This
version of the dataset, version {version}, contains information from
{n_pubs} publications, covering {n_interactions} pathogen--host
interactions and {n_pathogen_genes} pathogen genes across {n_pathogens}
pathogen species and {n_hosts} host species.
