<html>
<head>
<meta charset="utf-8">
<style>
body {font-family: sans-serif}
p, ol, ul {width: 40em}
table {border-collapse: collapse}
td, th {border: 1px solid black; padding: 0.2em 0.3em}
th {font-weight: bold}
</style>
</head>
<body>
<h1>The Pathogen-Host Interaction Database, version 4.12</h1>
<p><a href="https://doi.org/10.5281/zenodo.5356871">doi:10.5281/zenodo.5356871</a></p>
<ul>
<li><b>Version:</b> 4.12 (4.12.0)</li>
<li><b>Published:</b> 2021</li>
<li><b>Publisher:</b> PHI-base</li>
</ul>
<p><b>Cite as:</b> Carberry, J. S. (2021).
The Pathogen-Host Interactions Database, version 4.12 (v4.12)
[Data set]. Zenodo. <a href="https://doi.org/10.5281/zenodo.5356871">https://doi.org/10.5281/zenodo.5356871</a></p>
<h2>Description</h2>
<p>PHI-base is an online biological database (available at
<a href="http://www.phi-base.org">phi-base.org</a>) that catalogues experimentally
verified pathogenicity, virulence and effector genes from fungal,
oomycete and bacterial pathogens, which infect animal, plant, fungal and
insect hosts. PHI-base aims to aid discovery of genes in pathogens that
are of medical or agronomic importance, and which may be potential
targets for chemical intervention.</p>
<p>Each entry in PHI-base is curated by domain experts and is supported by
strong experimental evidence, as well as references to the literature in
which the original experiments are described. Each gene has a detailed
description of the predicted protein's function during the host
infection process. To facilitate data interoperability, we have
annotated genes using controlled vocabularies and links to external
sources (UniProt, Gene Ontology, Enzyme Commission, NCBI Taxonomy, EMBL,
PubMed and FRAC).</p>
<p>This PHI-base dataset is a Frictionless Data Package containing an
export of the PHI-base database in CSV format (comma-separated values).</p>
<p>Amino acid sequences for each gene in PHI-base are included as a 
supplementary FASTA file.</p>
<h2>Data contents</h2>
<table>
<thead>
<tr>
<th>File</th>
<th>Name</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>phi-base_4.12_data.csv</td>
<td>PHI-base 4.12 dataset</td>
<td>An export of data from PHI-base in CSV format.</td>
</tr>
<tr>
<td>phi-base_4.12_fasta.fas</td>
<td>PHI-base 4.12 FASTA file</td>
<td>Amino acid sequences for each gene in PHI-base, where available.</td>
</tr>
</tbody>
</table>
<h2>Authors</h2>
<table>
<thead>
<tr>
<th>Name</th>
<th>Email</th>
<th>ORCID ID</th>
<th>Affiliation</th>
<th>Role</th>
</tr>
</thead>
<tbody>
<tr>
<td>Josiah S. Carberry</td>
<td><a href="mailto:josiah.carberry@example.org">josiah.carberry@example.org</a></td>
<td><a href="https://orcid.org/0000-0002-1825-0097">0000-0002-1825-0097</a></td>
<td>Brown University</td>
<td>Principal Investigator</td>
</tr>
</tbody>
</table>
<h2>Contributors</h2>
<table>
<thead>
<tr>
<th>Name</th>
<th>ORCID ID</th>
<th>Affiliation</th>
<th>Role</th>
</tr>
</thead>
<tbody>
<tr>
<td>John Smith</td>
<td></td>
<td>Acme Corporation</td>
<td>Curator</td>
</tr>
<tr>
<td>Anonymous</td>
<td><a href="https://orcid.org/0000-0002-1825-0097">0000-0002-1825-0097</a></td>
<td></td>
<td>Lead curator</td>
</tr>
</tbody>
</table>
<p>The data content for PHI-base 4.12 was curated by curators at
<a href="https://molecularconnections.com/">Molecular Connections Pvt Ltd.</a>
in Bangalore, India.</p>
<p>(Note: for data protection reasons, the names and affiliations of
contributors cannot be included in the files within this dataset.
You can view this information on the <a href="https://doi.org/10.5281/zenodo.5356871">Zenodo record page</a>
for this dataset.)</p>
<h2>Conditions of use</h2>
<ul>
<li><b>Rights holder</b>: PHI-base</li>
<li><b>Licence</b>: This dataset is available under the <a href="https://creativecommons.org/licenses/by/4.0/">Creative Commons
    Attribution 4.0
    International</a>
    licence.</li>
<li><b>Cite as</b>: Carberry, J. S. (2021).
    The Pathogen-Host Interactions Database, version 4.12 (v4.12)
    [Data set]. Zenodo. <a href="https://doi.org/10.5281/zenodo.5356871">https://doi.org/10.5281/zenodo.5356871</a></li>
<li><b>Conditions of use</b>: PHI-base relies on the integrity of our users
    to ensure that PHI-base is acknowledged as the source of this 
    dataset. Citation enables us to monitor the use of this dataset and
    to demonstrate its value. Please send us a link to any publication
    that uses this dataset.</li>
</ul>
<h2>Funding</h2>
<p>PHI-base version 4.12 was funded by the <a href="http://dx.doi.org/10.13039/501100000268">Biotechnology and
Biological Sciences Research
Council</a> (BBSRC) through the
BBR award "A FAIR community resource for pathogens, hosts and their 
interactions to enhance global food security and human health"
[BB/S020020/1].</p>
<h2>Technical information</h2>
<p>This data package follows several standards that were created by the
<a href="https://frictionlessdata.io/">Frictionless Data</a> team:</p>
<ul>
<li>the <a href="https://specs.frictionlessdata.io/data-package/">Data Package</a>
    standard is used for the metadata of the entire dataset,</li>
<li>the <a href="https://specs.frictionlessdata.io/tabular-data-resource/">Tabular Data
    Resource</a>
    standard describes the PHI-base CSV dataset, and</li>
<li>the <a href="https://specs.frictionlessdata.io/table-schema/">Table Schema</a>
    standard describes the columns of the CSV file.</li>
</ul>
<p>Data in this dataset can be programmatically accessed using the
<a href="https://framework.frictionlessdata.io/">Frictionless Framework</a>, an
open-source framework written in Python.</p>
<h2>Data dictionary</h2>
<p>The following table describes the columns in the PHI-base CSV file
included with this data package. For a complete description with
examples and data validation rules, see the file
<b>phi-base_schema.json</b>.</p>
<table>
<thead>
<tr>
<th>Column</th>
<th>Name</th>
<th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>record_id</td>
<td>Record ID</td>
<td>string</td>
<td>The unique identifier for a record (row) in the PHI-base database.</td>
</tr>
<tr>
<td>phi_id</td>
<td>PHI ID</td>
<td>string</td>
<td>An identifier corresponding to one gene of one pathogen from one publication.</td>
</tr>
<tr>
<td>protein_id_source</td>
<td>Protein ID source</td>
<td>string</td>
<td>The source database for the protein identifier.</td>
</tr>
<tr>
<td>protein_id</td>
<td>Protein ID</td>
<td>string</td>
<td>The unique identifier for the protein expressed by the pathogen gene. The identifier references a protein database (such as UniProtKB).</td>
</tr>
<tr>
<td>gene_id_source</td>
<td>Gene ID source</td>
<td>string</td>
<td>The source database for the pathogen gene identifier.</td>
</tr>
<tr>
<td>gene_id</td>
<td>Gene ID</td>
<td>string</td>
<td>The unique identifier for the pathogen gene, which references a genetic database (such as GenBank).</td>
</tr>
<tr>
<td>sequence_strain</td>
<td>Sequence strain</td>
<td>string</td>
<td>The strain, isolate, subspecies, etc. of the pathogen for the gene sequence or protein sequence.</td>
</tr>
<tr>
<td>gene</td>
<td>Gene</td>
<td>string</td>
<td>The gene name for the pathogen gene.</td>
</tr>
<tr>
<td>chromosome_location</td>
<td>Chromosome location</td>
<td>string</td>
<td>The chromosome where the pathogen gene is located.</td>
</tr>
<tr>
<td>gene_protein_modification</td>
<td>Gene protein modification</td>
<td>string</td>
<td>Specific modifications to the targeted pathogen protein or promoter, such as amino acid modification, promoter sequence modification, or domain swaps.</td>
</tr>
<tr>
<td>modified_gene_protein_id</td>
<td>Modified gene protein ID</td>
<td>string</td>
<td>The unique identifier for the modified pathogen gene.</td>
</tr>
<tr>
<td>interacting_partners</td>
<td>Interacting partners</td>
<td>string</td>
<td>Proteins or DNA elements within the pathogen that interact with the pathogen gene. Multiple values are delimited with a semicolon.</td>
</tr>
<tr>
<td>interacting_partners_id</td>
<td>Interacting partners ID</td>
<td>string</td>
<td>The unique identifiers for the proteins or DNA elements within the pathogen that interact with the pathogen gene. Each identifier is prefixed with with the name of its source database. Multiple values are delimited with a semicolon.</td>
</tr>
<tr>
<td>multiple_mutation</td>
<td>Multiple mutation</td>
<td>string</td>
<td>One or more PHI IDs for pathogen genes that are disrupted as part of a mutation. Multiple values are delimited with a semicolon.</td>
</tr>
<tr>
<td>pathogen_id</td>
<td>Pathogen ID</td>
<td>integer</td>
<td>The NCBI Taxonomy identifier for the pathogen.</td>
</tr>
<tr>
<td>pathogen_species</td>
<td>Pathogen species</td>
<td>string</td>
<td>The scientific name of the pathogen.</td>
</tr>
<tr>
<td>pathogen_strain_id</td>
<td>Pathogen strain ID</td>
<td>integer</td>
<td>The NCBI Taxonomy identifier at the strain level for the strain of the pathogen.</td>
</tr>
<tr>
<td>pathogen_strain</td>
<td>Pathogen strain</td>
<td>string</td>
<td>The names of the strain, isolate, subspecies, pathovars, serotypes etc. of the pathogen. Multiple values are delimited with a semicolon.</td>
</tr>
<tr>
<td>disease</td>
<td>Disease</td>
<td>string</td>
<td>The names of any diseases known to be caused by the pathogen. Multiple values are delimited with a semicolon.</td>
</tr>
<tr>
<td>host_description</td>
<td>Host description</td>
<td>string</td>
<td>High-level taxonomic classification for the host species, such as an order or clade.</td>
</tr>
<tr>
<td>host_id</td>
<td>Host ID</td>
<td>integer</td>
<td>The NCBI Taxonomy identifier for the host organism.</td>
</tr>
<tr>
<td>host_species</td>
<td>Host species</td>
<td>string</td>
<td>The scientific name of the host organism. The common name of the species (if any) is included in parentheses after the scientific name.</td>
</tr>
<tr>
<td>host_strain</td>
<td>Host strain</td>
<td>string</td>
<td>The name of the strain, cell line, genotype, cultivar, ecotype etc. of the host species. Multiple values are delimited with a semicolon.</td>
</tr>
<tr>
<td>host_genotype</td>
<td>Host genotype</td>
<td>string</td>
<td>Any host genes of interest which are not necessarily tested, but referred to in the publication with regard to the pathogen gene of interest. Multiple values are delimited with a semicolon.</td>
</tr>
<tr>
<td>host_genotype_id</td>
<td>Host genotype ID</td>
<td>string</td>
<td>The unique identifiers for the host genes referred to in the publication. Each identifier is prefixed with with the name of its source database. Multiple values are delimited with a semicolon.</td>
</tr>
<tr>
<td>tissue</td>
<td>Tissue</td>
<td>string</td>
<td>The host tissue where the experiment was initiated. The resulting diseased tissue may be provided in parentheses. Most tissue names follow the naming conventions of the BRENDA Tissue Ontology (BTO). Multiple values are delimited with a semicolon.</td>
</tr>
<tr>
<td>gene_function</td>
<td>Gene function</td>
<td>string</td>
<td>The biological function of the disrupted pathogen gene.</td>
</tr>
<tr>
<td>go_annotation</td>
<td>Gene Ontology annotation</td>
<td>string</td>
<td>List of all appropriate Gene Ontology (GO) terms or Enzyme Commission (EC) numbers for the pathogen gene of interest. GO terms may be followed by a GO evidence code (e.g. IMP). Multiple values are delimited with a semicolon.</td>
</tr>
<tr>
<td>database</td>
<td>Database</td>
<td>string</td>
<td>The name of the source database for records in the 'GO annotation' column.</td>
</tr>
<tr>
<td>pathway</td>
<td>Pathway</td>
<td>string</td>
<td>The name of the pathway that the disrupted pathogen gene is involved in.</td>
</tr>
<tr>
<td>mutant_phenotype</td>
<td>Mutant phenotype</td>
<td>string</td>
<td>Broadly describes a phenotype that results from a pathogen-host interaction, or the response of a pathogen to some chemical.</td>
</tr>
<tr>
<td>mating_defect</td>
<td>Mating defect</td>
<td>string</td>
<td>Whether the experimental pathogen has a mating defect prior to penetrating the host (yes or no). Additional information may be included in parentheses.</td>
</tr>
<tr>
<td>pre_penetration_defect</td>
<td>Pre-penetration defect</td>
<td>string</td>
<td>Whether the experimental pathogen has a pre-penetration defect (yes or no). Additional information may be included in parentheses.</td>
</tr>
<tr>
<td>penetration_defect</td>
<td>Penetration defect</td>
<td>string</td>
<td>Whether the experimental pathogen has a penetration defect (yes or no). Additional information may be included in parentheses.</td>
</tr>
<tr>
<td>post_penetration_defect</td>
<td>Post-penetration defect</td>
<td>string</td>
<td>Whether the experimental pathogen has a post-penetration defect (yes or no). Additional information may be included in parentheses.</td>
</tr>
<tr>
<td>disease_manifestation</td>
<td>Disease manifestation</td>
<td>string</td>
<td>Description of how disease develops during the pathogen-host interaction (at a macroscopic scale).</td>
</tr>
<tr>
<td>vegetative_spores</td>
<td>Vegetative spores</td>
<td>string</td>
<td>Description of the experimental pathogen's vegetative spores.</td>
</tr>
<tr>
<td>sexual_spores</td>
<td>Sexual spores</td>
<td>string</td>
<td>Description of the experimental pathogen's sexual spores.</td>
</tr>
<tr>
<td>in_vitro_growth</td>
<td>In vitro growth</td>
<td>string</td>
<td>Description of how the experimental pathogen grows in vitro.</td>
</tr>
<tr>
<td>spore_germination</td>
<td>Spore germination</td>
<td>string</td>
<td>Description of the experimental pathogen's spore germination process.</td>
</tr>
<tr>
<td>essential_gene</td>
<td>Essential gene</td>
<td>string</td>
<td>Whether a knockout of the pathogen gene is lethal to the pathogen; that is, whether the gene is essential for the in vitro or in vivo growth of the pathogen (yes or no).</td>
</tr>
<tr>
<td>gene_inducer</td>
<td>Gene inducer</td>
<td>string</td>
<td>Chemical substance that can alter metabolic reactions within a cell, tissue, or intact host organism. Multiple values are delimited with a semicolon.</td>
</tr>
<tr>
<td>gene_inducer_id</td>
<td>Gene inducer ID</td>
<td>string</td>
<td>Accession numbers for the inducer chemicals, sourced from ChEBI or the CAS Registry. Accession numbers for each chemical are separated by a semicolon. Where there are multiple accessions for one chemical, the accession numbers are separated by a comma.</td>
</tr>
<tr>
<td>host_target</td>
<td>Host target</td>
<td>string</td>
<td>The first-host-target proteins which interact with the corresponding pathogen protein. May include the host cellular location where the experiment was observed. Multiple values are delimited with a semicolon.</td>
</tr>
<tr>
<td>host_target_id</td>
<td>Host target ID</td>
<td>string</td>
<td>Accession numbers for the host target genes of interest. Each accession number is prefixed with with the name of its source database. Multiple values are delimited with a semicolon.</td>
</tr>
<tr>
<td>interaction_phenotype</td>
<td>Interaction phenotype</td>
<td>string</td>
<td>Description of the interaction phenotype between the pathogen and the host organism.</td>
</tr>
<tr>
<td>host_response</td>
<td>Host response</td>
<td>string</td>
<td>The response observed in the host organism when it interacts with the altered pathogen gene (or protein) of interest.</td>
</tr>
<tr>
<td>exp_technique_stable</td>
<td>Experimental technique (stable)</td>
<td>string</td>
<td>Description of the experimental procedure.</td>
</tr>
<tr>
<td>exp_technique_transient</td>
<td>Experimental technique (transient)</td>
<td>string</td>
<td>Describes the experimental evidence for a transient assay. Used when the experimental technique is 'functional test in host: transient expression' or 'other evidence'.</td>
</tr>
<tr>
<td>pmid</td>
<td>PubMed ID</td>
<td>integer</td>
<td>The PubMed identifier (PMID) for the referenced publication, without the 'PMID:' prefix.</td>
</tr>
<tr>
<td>reference_source</td>
<td>Reference source</td>
<td>string</td>
<td>The name of the library or information resource that contains the referenced publication.</td>
</tr>
<tr>
<td>doi</td>
<td>DOI</td>
<td>string</td>
<td>The Digital Object Identifier for the referenced publication.</td>
</tr>
<tr>
<td>reference_detail</td>
<td>Reference detail</td>
<td>string</td>
<td>The full citation of the referenced publication. May be used as a substitute for a missing PubMed ID or DOI.</td>
</tr>
<tr>
<td>comments</td>
<td>Comments</td>
<td>string</td>
<td>Additional information, e.g. information found only in the abstract of the referenced publication, or information that is not compatible with any other column in the PHI-base database.</td>
</tr>
<tr>
<td>author_reference</td>
<td>Author reference</td>
<td>string</td>
<td>An abbreviated citation for the referenced publication.</td>
</tr>
<tr>
<td>year</td>
<td>Year</td>
<td>year</td>
<td>The year of publication of the referenced publication.</td>
</tr>
<tr>
<td>curation_date</td>
<td>Curation date</td>
<td>date</td>
<td>The date when the referenced publication was curated.</td>
</tr>
</tbody>
</table>
</body>
</html>
